go/hyperkit.go

```diff
@@ -131,6 +131,9 @@ type HyperKit struct {
 	// process died unexpectedly.
 	ExtraFiles []*os.File `json:"extra_files"`
 
+	// ErrorCh is where the result of Wait'ing for hyperkit is sent.
+	ErrorCh chan error `json:"-"`
```

Comment:
I wonder if this might better fit as something returned by `Start`?

Reply:
Yep, I considered this too.  I chose to make it a member to avoid changing the signature (for backward compatibility with the code that already uses Start), but also because it allows to avoid having too many return values.  Finally, at some point I considered being able to provide `Start` with the channel I want it to uses (it could simplify a tiny bit our use in D4M), but really that's hardly a valid argument.

I can do whatever you prefer about this.

Reply:
I think, but I'm not sure, that the assignment to `h.ErrorCh` in `FromState` is wrong, since in that case there is nothing to feed the error code into the channel -- you wouldn't call `Start` or `Run` under these circumstances since the process is considered to already be running.

If I'm right then rather than dealing with the question of when exactly `h.ErrorCh` will be non-nil and meaningful I think it would be better to make the channel be part of the return from `Start`. That way only callers of `Start` can have any expectation of having a valid error channel.

It also makes the ownership of the channel clearer, rather than any code which happens to have an `h` in its hand potentially thinking it is allowed to receive from `h.ErrorCh`.

Reply:
You are right.

Actually the whole point of FromState is unclear to me: what is it used for?

There's one final advantage that we had for putting the channel in the struct, it's that we are considering running several VMs, in which case it was easier that each vm had its own self-contained structure, rather than having to deal with a slice of structures and a slice of channels.

Reply:
`FromState` is used to recreate a HyperkIt object from the state directory. So you could start it in the background and then sometime later, in a different process can re-create the object to manage it.

Reply:
@rn Seems like we commented in parallel! (https://github.com/moby/hyperkit/pull/174#issuecomment-356299316). Is that functionality actually used anywhere?

The wrinkle here is that a object which is "recovered" in this way using `FromState` cannot expose the exit code of the process (since it is not performing the wait on it etc and hyperkit may not even be its child).